File: src/main/java/algo/prefixSum/ShortestSubarrayWithSumAtLeastK_862.java
package algo.prefixSum;

import java.util.ArrayDeque;

public class ShortestSubarrayWithSumAtLeastK_862 {
  public static int shortestSubarray(int[] nums, int k) {
    var len = nums.length;
    var q = new ArrayDeque<Integer>();  // increasing queue
    var sum = new long[len + 1];  // prefix sum at i
    for (var i = 0; i < len; i++) {
      sum[i + 1] = sum[i] + nums[i];
    }

    var shortest = Integer.MAX_VALUE;
    for (var i = 0; i <= len; i++) {
      while (!q.isEmpty() && sum[i] < sum[q.getLast()]) {
        q.pollLast();
      }
      q.offerLast(i);
      while (!q.isEmpty() && sum[i] - sum[q.getFirst()] >= k) {
        shortest = Math.min(shortest, i - q.getFirst());
        q.pollFirst();
      }
    }
    return shortest != Integer.MAX_VALUE ? shortest : -1;
  }

  public static void main(String[] args) {
    System.out.println(shortestSubarray(new int[]{1}, 1));
    System.out.println(shortestSubarray(new int[]{1,2}, 4));
    System.out.println(shortestSubarray(new int[]{2,-1,2}, 3));
    System.out.println(shortestSubarray(new int[]{17,85,93,-45,-21}, 150));
    System.out.println(shortestSubarray(new int[]{56,-21,56,35,-9}, 61));
    System.out.println(shortestSubarray(new int[]{-28,81,-20,28,-29}, 89));
    System.out.println(shortestSubarray(new int[]{-34,37,51,3,-12,-50,51,100,-47,99,34,14,-13,89,31,-14,-44,23,-38,6}, 151));
  }
}

File: src/main/kotlin/algo/prefixSum/ShortestSubarrayWithSumAtLeastK_862.kt
package algo.prefixSum

class KtShortestSubarrayWithSumAtLeastK_862 {
    fun shortestSubarray(nums: IntArray, k: Int): Int {
        val prefixSum = LongArray(nums.size + 1)
        nums.indices.forEach { idx ->
            prefixSum[idx + 1] = nums[idx] + prefixSum[idx]
        }

        // maintain an increasing sequence of prefix sum and check the diff between two sums
        val maxQ = ArrayDeque<Int>()
        var minLen = Int.MAX_VALUE
        prefixSum.indices.forEach { idx ->
            val value = prefixSum[idx]
            while (maxQ.isNotEmpty() && prefixSum[maxQ.last()] > value) {
                maxQ.removeLast()
            }
            maxQ.addLast(idx)
            while (maxQ.isNotEmpty() && value - prefixSum[maxQ.first()] >= k) {
                val e = maxQ.removeFirst()
                minLen = minLen.coerceAtMost(idx - e)
            }
        }
        return if (minLen == Int.MAX_VALUE) -1 else minLen
    }
}

fun main() {
    val subarray = KtShortestSubarrayWithSumAtLeastK_862()
    println(subarray.shortestSubarray(intArrayOf(1), 1))
    println(subarray.shortestSubarray(intArrayOf(1,2), 4))
    println(subarray.shortestSubarray(intArrayOf(2,-1,2), 3))
}